### DISCUSSION TITLE
Major changes in BSB_LAN_defs.h

### DISCUSSION THREAD

**Discussion by fredlcore** [fredlcore/BSB-LAN#482] [General]
**TLDR: We will no longer provide updated BSB_LAN_defs.h after version 2.0 for safety reasons. Users will get tailor-made BSB_LAN_defs.h for their systems.**

Some of you have probably wondered why there haven't been any significant updates in BSB_LAN_defs.h despite people sending in their "Check for new parameter" reports. In fact, these keep piling up on my ToDo list, and there are probably over a 1000 parameters waiting to be included.

Apart from the fact that I had quite a bumpy ride in life over the last 12 months, the reason for this delay is rather caused by the fact that we have managed to find a way to figure out what parameters each heating system "knows". That is the "complete dump" that is part of the "Check for new parameters" for quite some time now. 

At first we thought that this is great - we would no longer have to ask users to check all different kind of parameters and their options and could simply add any parameter that is not currently in our BSB_LAN_defs.h. But after analyzing more and more of these reports, we stumbled upon two major problems:

The first one was that over the 20 years or so that the LPB/BSB bus exists, Siemens (the company manufacturing all the controllers in the heating systems BSB-LAN supports) had to introduce more and more categories and parameters that eventually overlapped with other, already existing categories or parameters. So one category or group of parameters on one system has the same number range as a different category or group of parameters has on another system. We have tried to accomodate this so far by moving these numbers away to "free" locations, but these became fewer and fewer. It also made translations more difficult because you would have to check whether parameter X on machine A really means the same on machine B or not. Also, we would have to change the whole category system which currently is a "one fits all" when in fact it's not. While all this would be possible to deal with, this would have meant a lot of work for us because we would have to do this check with every new heating system we come across.

The second and much more severe problem is that we figured out that there are a lot of parameters which have the same parameter number and the same name, but come with completely different sets of options. When you look at parameter 5950, you see that we realized this already years back but always hoped that this was just a strange exception. The fact is: it's not. There are dozens and dozens of parameters that differ from one heating system to another, be it in terms of their options, their conversion factor, their unit etc. We even realized that some options have the same numeric values, but with an opposite meaning (think: inverted contact function). And this is where we had to think hard, because in the current system, we would always assume a "default" meaning without knowing whether there is actually such a thing as a "default" meaning of a parameter. Take again parameter 5950 for example (there are many, many more): We think that the DEV_ALL line is the default that we use for any unspecified or new machine when in fact the list of exceptions that follows is so long that it's probably the other way around. This becomes worse when you compare different heating systems, such as gas heaters and heat pumps for example. We started out with gas and oil burners in the beginning, but again, there are so many more different systems now that it's hard to keep track.

So unless users really check if the parameters and options shown in BSB-LAN actually match what is shown on the heater, users might just take this as a given and use the parameter unknowingly in the wrong way. The danger here is that the heater will accept the "wrong" setting because the telegram that BSB-LAN sends is valid - just the meaning might be more or less completely off. And there are dozens of parameters, especially those with drop-down menus (a.k.a. ENUM parameters) which are f..cked up beyond repair.

So where do we go from here? We have contemplated various options but the only one that seems feasible is to move away from the "all parameters for all heating systems" approach. There was very little gain for the average user to begin with, but with the abovementioned risks, it is simply no responsible approach anymore.
The good news is that we now have more or less decoded the complete dump so we can generate a tailor-made BSB_LAN_defs.h from this dump for the heating system in question with all the correct parameter names, options, units etc.
The not so good news is that this process can only be semi-automated. It will require manual adjustments by us. And before anyone asks: No, we cannot make public the decoding algorhythm of the complete dump. Legal regulations in Germany do allow reverse engineering to some extent, but not to publish the decoding algorhythm.

So this is how we plan to do this in the near future:
- People currently using version 2.0 can still do so as long as they want. However, we strongly recommend to check each and every parameter thoroughly whether it actually works on your heating system the same way as BSB-LAN says that it does.
- Users who want to use the device specific BSB_LAN_defs.h can send in their "Check for new parameters" (we will rename that function because it won't do what it says anymore) and will get a device specific BSB_LAN_defs.h in return.
- For new users, BSB-LAN will be delivered with a minimal set of parameters that are proven to work across all systems to test the functionality. They can either take the BSB_LAN_defs.h from version 2.0 if they really want to, or they can also send us the log and get the device specific BSB_LAN_defs.h.
- All users can manually add parameters they know to be working from the current 2.0 BSB_LAN_defs.h. So if your heater does not advertise a command that you know is working correctly, you can simply add it. The BSB_LAN_defs.h will not be updated when you update the main code.

We know that this is more work on all sides (more on ours than on the users'), but we believe that having a fully and accurately working BSB-LAN is in everyone's best interest. The old system was nice as a playground and especially I loved to explore new parameters, but a heating system is not a playground, therefore these changes are necessary.

We will roll out these changes over the next couple of weeks, hopefully before the heating season really kicks in, but bear with us if it'll take a bit longer. We want to make this step well prepared and without any major obstacles for the users.

More information will be provided here whenever there is any.

**Comment by DE-cr** [comment]
"**Security** reasons"? => "**Safety** stands for accident avoidance, and **security** for crime prevention." (Source: https://www.tuev-nord.de/explore/en/explains/whats-the-difference-between-safety-and-security/) ;)

**Reply by fredlcore** [reply]
Yes, then safety is the better term, edited accordingly above.

**Comment by fredlcore** [comment]
Since people have asked about this in some of the pull requests: The parameter descriptions can be extracted in a number of languages, so translating parameters will no longer be an issue. Users will just have to state which language they want their BSB_LAN_custom_defs.h to be in.

**Comment by fredlcore** [comment]
![0942475A-AB0B-418C-9748-E5FF95126D52_1_105_c](https://user-images.githubusercontent.com/24660955/196021851-e02eb3a8-8bbd-481b-9857-6be09cadb09c.jpeg)

**Reply by fredlcore** [reply]
Just to show you the amount of manual work involved in translating the dump into the actual BSB_LAN_custom_defs.h using a QAA75...

**Comment by Xeevis** [comment]
> Users who want to use the device specific BSB_LAN_defs.h can send in their "Check for new parameters" (we will rename that function because it won't do what it says anymore) and will get a device specific BSB_LAN_defs.h in return.

@fredlcore Thanks for all the hard work, BSB-LAN has been working flawlessly for me since 2023 🥳. Just a question, does this device specific file get any updates? I'm holding onto one from 2023 and running BSB-LAN 3.2.2 without any problems, but I'm doing update pass on all my devices and would like to update to all the latest. Do I just go with my old defs with 5.0? Thanks.

**Reply by fredlcore** [reply]
No, the file stays the same because it comes from your controller. So only if you get a new controller, you need a new list. Otherwise, you can reuse the original one at all times. After skipping several version numbers, you might need to adjust the semantics here and there, similar to the configuration file, so check the ChangeLog for that.

**Comment by Lamarqe** [comment]
> Users who want to use the device specific BSB_LAN_defs.h can send in their "Check for new parameters" (we will rename that function because it won't do what it says anymore) and will get a device specific BSB_LAN_defs.h in return.

Where can I find this option? I would like to receive the device specific BSB_LAN_defs.h for my heater (Fujitsu Waterstage WSYA100DD6).

Note: I am not using BSB-LAN directly. 
Instead, I use [bsbgateway](https://github.com/loehnertj/bsbgateway) and intend to use the header file to "convert" it in my personal version of bsbgateway.

**Reply by fredlcore** [reply]
If you don't use BSB-LAN, then, obviously, there is no such option. bsbgateway is a completely different, and unrelated, software.

**Reply by Lamarqe** [reply]
Thanks for the super fast response! Yes, I understand. 

Could you please still send me the device specific header file for the mentioned model above?
It is absolutely clear that I can not use the information as-is.

I intend to extract the device configuration from the header file and with that re-create the device specific configuration in bsbgateway.

**Reply by fredlcore** [reply]
No, I can't. Please read and understand the process described above. Without the dump file, I cannot generate any of those lists.

**Reply by Lamarqe** [reply]
Is it correct that the "Check for new parameters" option sends certain BSB-messages to the heater which identify the specifics of the device?

If yes, is it possible to share which messages it sends? 
I could then generate the same messages and collect their responses via bsbgateway, then share them with you and finally you could use them to generate the device specific header file?

**Reply by fredlcore** [reply]
Yes, it does. Have a look at the source code if you are so hell-bent on avoiding BSB-LAN, and want to do it your way, but don't expect me to put in a significant amount of time to explain the process.

**Reply by Lamarqe** [reply]
Okay, that a sufficiently dismissive response for me to learn that this is not desired.
Sorry, I really did expect this reaction. 
I did not expect that you are strongly motivated that people use bsb-lan and not alternatives.

But no problem, I can understand. I will see what I can do. 
Thank you.

**Reply by fredlcore** [reply]
I'm not strongly motivated that people use BSB-LAN instead of alternatives. On the contrary, I have been in good contact with Johannes, the author of bsbgateway. I also recommend his software to people who for some reasons prefer to use a Raspberry instead of an ESP32, and I have even adjusted by BSB-LAN adapters so that they also work with his software. But it is clear to everyone involved that bsbgateway has a much more limited scope than BSB-LAN, so users generally know why they choose either of these solutions.

You were expecting me to explain a very complex system and assume that you are capable enough to implement that in bsbgateway. If that's the case, then my suggestion to have a look at my source code would be sufficient for you to do that. If you rather expect me to explain every little detail about this process (which, as explained above, I cannot legally do publicly), and then probably answer a lot of follow-up questions in case you don't get it done right the first time, then you are assuming that I have a lot of free time at my hands (which I don't). I invest my free time in supporting BSB-LAN users and working on the BSB-LAN project. If you really want to save a few Euros for an ESP32 which would allow you to simply extract the dump yourself and then do with it whatever you want, then you'll have your good reasons for that. But I hope you can understand that those might not be good reasons from my point of view.

**Reply by Lamarqe** [reply]
I use bsbgateway because the solution fits very well into my environment, it has extremely low complexity and worked for me (read-access) immediately when I tried it 3 years ago. Just like plug&play.
I also want to mention that it was BSB-LAN which pointed me to this solution (I found it via the readme note). So, getting told that I would be stubborn is inadequate.

Now I was looking into getting the write access to work, tried the sniffer option and concluded that testing out all 50 write commands with all options would be some hours or even days of effort. Johannes' recent readme note about _the BSB_LAN project did an incredible job of gathering this information for hundreds of devices_ brought me here, asking if this information could be shared.
As the Waterstage WSYA* devices are explicitly listed in the supported devices, I was hoping that this device specific version had been created in the past and could be shared from historic sources. The question about the internals was purely to build up more context for me. The idea to reproduce messages via bsbgateway came to me after 10 seconds of thinking

I started looking into the source code for some minutes, but it is not as straight-forward to me as bsbgateway.
Furthermore, analyzing source code which cannot be executed without getting the additional hardware is not easy to do.

So in other words, if I need to debug and understand the implementation of BSB-LAN, the idea to use the BSB-LAN header as starting point for some bsbgatway adaptions is inferior to either simply sniffing the options in a some evening hours or as you suggested to invest into an ESP32 (the main hurdle here is not the cost. Its the entry hurdle to the ESP eco-system (never used it before) and the (in my personal view) added complexity of the solution.

**Reply by fredlcore** [reply]
Yes, we recommend and respect each other's projects. However, the scope is different. Wanting something from one project that is part of the other one simply doesn't work. I thought that I made this clear earlier, but in case I didn't, then I'm sorry for assuming.

And yes, the Waterstage is supported. But there can be dozens of controller types and/or firmware versions in that model range which results in dozens or even hundreds of different parameters. That's what the text at the top tries to explain. I therefore cannot just generate a file for you even if I take previous submissions into account. Yours could be an old one or a new one, we don't know until we know. 

I suggest you follow the path that takes the least time for you. If you think that BSB-LAN (without having even tried it) is not as straight-forward as bsbgateway and therefore it's easier to sniff the options by yourself, then that's the solution you should go for.

**Reply by Lamarqe** [reply]
Yep, we are aligned.
Thanks once more for taking the time for this detailed exchange. 

And Kudos for creating this highly successful project 👍
